---
title: "diff118_iNiDA_d23_CtrlSMPD1ASAH1"
output: html_document
date: "2024-10-10"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# load packackes & set project directory
```{r}
# --- Load Required Libraries ---
# --- Core Tidyverse ---
library(tidyverse)     # includes dplyr, ggplot2, tidyr, purrr, tibble, readr, stringr, etc.

# --- Plotting Enhancements ---
library(ggdist)        # stat_halfeye
library(gghalves)      # geom_half_point
library(ggrepel)       # text labels (if used)
library(pheatmap)      # for heatmaps
library(RColorBrewer)  # custom color scales
library(viridis)       # color scales (optional)
library(NatParksPalettes)
library(scales) 

# --- Plot Layouts ---
library(gridExtra)     # arranging plots
library(cowplot)       # aligning and combining plots
library(patchwork)     # alternate plot composition syntax


# --- Data Reshaping and Manipulation ---
library(reshape2)      # melt, dcast (legacy but used in some places)
library(forcats)       # factor reordering
library(data.table)    # fast table handling (if used)

library(lintr)
library(devtools)

```


# output dirs overview:
```{r}
out_dir_diff118_dataframes <- "/Users/felix/Dropbox (HMS)/Felix/Harvard/03_LSD-PD/Proteomics/20240801_dif118_iN-iDA_Ctrl-SMPD1-ASAH1_Proteomics/Eval/dataframes"
dir.create(out_dir_diff118_dataframes, recursive = TRUE, showWarnings = FALSE)

out_dir_diff118_heatmap  <- "/Users/felix/Dropbox (HMS)/Felix/Harvard/03_LSD-PD/Proteomics/20240801_dif118_iN-iDA_Ctrl-SMPD1-ASAH1_Proteomics/Eval/heatmaps"
dir.create(out_dir_diff118_heatmap, recursive = TRUE, showWarnings = FALSE)

out_dir_diff118_corrplots <- "/Users/felix/Dropbox (HMS)/Felix/Harvard/03_LSD-PD/Proteomics/20240801_dif118_iN-iDA_Ctrl-SMPD1-ASAH1_Proteomics/Eval/Organelle_CrossCorr"
dir.create(out_dir_diff118_corrplots, recursive = TRUE, showWarnings = FALSE)

```



# I. Background diff 136 axonal proteome Ctrl / ASAH1 e1 day 35 iN differentiation 
Dataset contains in vitro differentiaed iNeurons and iDA at day 23 of differentiation
Genotypes: H9 NGN2 TMEM192-3xHA (het) Control and ASAH1-/- clone E1 and SMPD1-/- clone E8

Samples (all in triplicates, all whole-cell (WC): Ctrl iN, Ctrl iDA, ASAH1 iN, ASAH1 iDA, SMPD1 iN, SMPD1 iDA
18plex_TMTpro & run on Thermo Orbitrap Eclipse w/FAIMS-Pro, hrMS2
data was created with MSstats, has p/q.values and log2FC for samples 
k-means clustering was performed on data, is in column "cluster"



# II. read in data & add protein organelle annotations 

```{r}
# ─────────────────────────────────────────────────────────────────────────────
# 1.  read the data  
# ─────────────────────────────────────────────────────────────────────────────

diff118_df <- read.csv('/Users/felix/Dropbox (HMS)/Felix/Harvard/03_LSD-PD/Proteomics/20240801_dif118_iN-iDA_Ctrl-SMPD1-ASAH1_Proteomics/Eval/diff118_All_results_stats_localizationAdded.csv', header=T, sep=',', stringsAsFactors = FALSE)


# ─────────────────────────────────────────────────────────────────────────────
# 2.  Transform p/q-values to -log10, and filter non-finite log2FCs
# ─────────────────────────────────────────────────────────────────────────────

log2fc_cols <- c(
"log2FC_Ctrl_d23_iDA.Ctrl_d23_iN",
"log2FC_SMPD1_d23_iN.Ctrl_d23_iN",
"log2FC_SMPD1_d23_iDA.Ctrl_d23_iN",
"log2FC_ASAH1_d23_iN.Ctrl_d23_iN",
"log2FC_ASAH1_d23_iDA.Ctrl_d23_iN",
"log2FC_SMPD1_d23_iN.Ctrl_d23_iDA",
"log2FC_SMPD1_d23_iDA.Ctrl_d23_iDA",
"log2FC_ASAH1_d23_iN.Ctrl_d23_iDA",
"log2FC_ASAH1_d23_iDA.Ctrl_d23_iDA",
"log2FC_SMPD1_d23_iDA.SMPD1_d23_iN",
"log2FC_ASAH1_d23_iN.SMPD1_d23_iN",
"log2FC_ASAH1_d23_iDA.SMPD1_d23_iN",
"log2FC_ASAH1_d23_iN.SMPD1_d23_iDA",
"log2FC_ASAH1_d23_iDA.SMPD1_d23_iDA",
"log2FC_ASAH1_d23_iDA.ASAH1_d23_iN" 
)

# transform p / q values into -log10 values 
diff118_df <- diff118_df %>%
  mutate(across(contains("p.val"), ~ ifelse(. > 0, -log10(.), NA), .names = "neg_log10_{.col}"),
         across(contains("q.val"), ~ ifelse(. > 0, -log10(.), NA), .names = "neg_log10_{.col}")) %>%
  filter(if_all(all_of(log2fc_cols), ~ is.finite(.)))

# ─────────────────────────────────────────────────────────────────────────────
# 3.  Add Subcellular Annotation
# ─────────────────────────────────────────────────────────────────────────────

# --- Step 1: Define contaminant list (optional, can be skipped if not relevant)
contaminant_genes <- c(
  "KRT8", "KRT19", "KRT9", "KRT10", "KRT2", "KRT18", "KRT1", "KRT5",
  "COL10A1", "COL13A1", "COL14A1", "COL18A1", "COL1A1", "COL1A2",
  "COL26A1", "COL2A1", "COL4A1", "COL4A2", "COL5A1", "COL6A1", "COL6A2"
)

# --- Step 2: Load and process subcellular annotation
cols_to_read <- c(1,2,3,4,5,6,7,9,10,12,15,16,17,18,19,20,21,22,23,25,26,27,28,29,30,31,32)
subcell_df <- read_csv("/Users/felix/Dropbox (HMS)/Felix/Harvard/03_LSD-PD/neuroLSDgit/proteome/SubCellAnnotation.csv", show_col_types = FALSE)[, cols_to_read]
subcell_df[] <- lapply(subcell_df, as.character)

long_subcell_df <- subcell_df %>%
  pivot_longer(cols = everything(), names_to = "Localization", values_to = "Genes") %>%
  filter(!is.na(Genes) & Genes != "") %>%
  mutate(Genes = trimws(Genes)) %>%
  distinct()

diff118_df_binary_matrix <- long_subcell_df %>%
  mutate(value = TRUE) %>%
  pivot_wider(
    names_from = Localization,
    values_from = value,
    values_fill = list(value = FALSE)
  )

# --- Step 3: Merge with diff136_df by Gene.Symbol
diff118_df_annotated <- diff118_df %>%
  filter(!Gene.Symbol %in% contaminant_genes) %>%
  left_join(diff118_df_binary_matrix, by = c("Gene.Symbol" = "Genes"))

diff118_df_annotated <- diff118_df_annotated %>%
  mutate(across(where(is.logical) | where(is.numeric), ~ replace_na(., 0)))

# --- Step 4. save as csv file 
write.csv(diff118_df_annotated, 
          file = file.path(out_dir_diff118_dataframes,"diff118_iN-iDA_d23_Ctrl-ASAH1-SMPD1_wholecellproteome_annotated.csv"), row.names = FALSE)
```


### IIa. Heatmap (z-norm rows, cluster by cluster)
```{r}
# ─────────────────────────────────────────────────────────────────────────────
# Heatmap (z-norm rows, cluster by cluster)
# ─────────────────────────────────────────────────────────────────────────────

# --- 1. set annotations and sample names 
expr_cols <- c(
  "Ctrl_d23_iN_1", "Ctrl_d23_iN_2","Ctrl_d23_iN_3",
  "Ctrl_d23_iDA_1","Ctrl_d23_iDA_2","Ctrl_d23_iDA_3",
  "SMPD1_d23_iN_1","SMPD1_d23_iN_2","SMPD1_d23_iN_3",
  "SMPD1_d23_iDA_1","SMPD1_d23_iDA_2","SMPD1_d23_iDA_3",                          
  "ASAH1_d23_iN_1","ASAH1_d23_iN_2","ASAH1_d23_iN_3",
  "ASAH1_d23_iDA_1","ASAH1_d23_iDA_2","ASAH1_d23_iDA_3"
)

# Matrix from expression columns
mat <- diff118_df %>%
  filter(!is.na(Gene.Symbol)) %>%
  mutate(Gene.Symbol = make.unique(as.character(Gene.Symbol))) %>%
  select(Gene.Symbol, cluster, all_of(expr_cols))

# Store and remove cluster for sorting
row_cluster <- mat$cluster
rownames(mat) <- mat$Gene.Symbol
mat <- mat[, expr_cols]

# Reorder rows by cluster
mat <- mat[order(row_cluster), , drop = FALSE]


# --- 2. set cluster colors and cluster annotations as rows
# Define full cluster levels explicitly
cluster_levels <- 0:8

# Force 'cluster' to be a factor with all levels, even if not present
row_annotation <- data.frame(
  cluster = factor(row_cluster[order(row_cluster)], levels = cluster_levels)
)
rownames(row_annotation) <- rownames(mat)

# Assign distinct colors
cluster_palette <- setNames(
  RColorBrewer::brewer.pal(9, "Pastel1"),
  as.character(cluster_levels)
)

# Add to annotation color list
ann_colors <- list(
  Genotype = c(Ctrl = "grey40", ASAH1 = "dodgerblue3", SMPD1 = "tomato3"),
  Fraction = c(`iN` = "#f1eef6",
               iDA  = "#74a9cf")
)

ann_colors$cluster <- cluster_palette


# --- 3. column annotations
annotation_col <- tibble(Sample = expr_cols) %>%
  mutate(
    Genotype = case_when(
      str_detect(Sample, "^Ctrl")   ~ "Ctrl",
      str_detect(Sample, "^ASAH1")  ~ "ASAH1",
      str_detect(Sample, "^SMPD1")  ~ "SMPD1",
      TRUE                          ~ NA_character_
    ),
    Fraction = case_when(
      str_detect(Sample, "_iN_")    ~ "iN",
      str_detect(Sample, "_iDA_")   ~ "iDA",
      TRUE                          ~ NA_character_
    )
  ) %>%
  column_to_rownames("Sample")


# --- 4. plot heatmap, save as pdf and store underlying dataframe as csv 
pdf_file <- file.path(out_dir_diff118_heatmap, "diff118_iN-iDAd35_Ctrl-ASAH1-SMPD1_heatmap.pdf")
cairo_pdf(pdf_file, width = 9, height = 10)

pheatmap(
  mat,
  scale              = "row",
  color              = colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(100),
  annotation_col     = annotation_col,
  annotation_row     = row_annotation,
  annotation_colors  = ann_colors,
  breaks             = seq(-4, 4, length.out = 101),
  cluster_rows       = FALSE,
  cluster_cols       = TRUE,
  show_rownames      = FALSE,
  show_colnames      = TRUE,
  angle_col          = 90,
  main               = "diff118 day 23 iN & iDA whole cell proteome: Ctrl, ASAH1, SMPD1",
  fontsize           = 6,
  border_color       = NA
)

dev.off()

# Add cluster as first column to expression matrix
mat_with_cluster <- cbind(cluster = row_annotation$cluster, mat)

# Write to CSV with cluster
write.csv(mat_with_cluster,
          file = file.path(out_dir_diff118_heatmap,"diff118_iN-iDAd35_Ctrl-ASAH1-SMPD1_heatmap.csv"),
          row.names = TRUE)
```


### IIb. Heatmap KO/Ctrl (still keep z-norm rows, cluster by cluster)
```{r}
# ─────────────────────────────────────────────────────────────────────────────
# Heatmap (z-norm rows, cluster by cluster)
# ─────────────────────────────────────────────────────────────────────────────

# --- 1. set annotations and sample names 
expr_cols <- c(
"log2FC_SMPD1_d23_iN.Ctrl_d23_iN","log2FC_ASAH1_d23_iN.Ctrl_d23_iN",
"log2FC_ASAH1_d23_iDA.Ctrl_d23_iDA","log2FC_SMPD1_d23_iDA.Ctrl_d23_iDA"                
)

# Matrix from expression columns
mat_norm <- diff118_df %>%
  filter(!is.na(Gene.Symbol)) %>%
  mutate(Gene.Symbol = make.unique(as.character(Gene.Symbol))) %>%
  select(Gene.Symbol, cluster, all_of(expr_cols))

# Store and remove cluster for sorting
row_cluster <- mat_norm$cluster
rownames(mat_norm) <- mat_norm$Gene.Symbol
mat_norm <- mat_norm[, expr_cols]

# Reorder rows by cluster
mat_norm <- mat_norm[order(row_cluster), , drop = FALSE]


# --- 2. set cluster colors and cluster annotations as rows
# Define full cluster levels explicitly
cluster_levels <- 0:8

# Force 'cluster' to be a factor with all levels, even if not present
row_annotation <- data.frame(
  cluster = factor(row_cluster[order(row_cluster)], levels = cluster_levels)
)
rownames(row_annotation) <- rownames(mat)

# Assign distinct colors
cluster_palette <- setNames(
  RColorBrewer::brewer.pal(9, "Pastel1"),
  as.character(cluster_levels)
)

# Add to annotation color list
ann_colors <- list(
  Genotype = c(Ctrl = "grey40", ASAH1 = "dodgerblue3", SMPD1 = "tomato3"),
  Fraction = c(`iN` = "#f1eef6",
               iDA  = "#74a9cf")
)

ann_colors$cluster <- cluster_palette


# --- 3. column annotations
annotation_col <- tibble(Sample = expr_cols) %>%
  mutate(
    Genotype = case_when(
      str_detect(Sample, "^Ctrl")   ~ "Ctrl",
      str_detect(Sample, "^ASAH1")  ~ "ASAH1",
      str_detect(Sample, "^SMPD1")  ~ "SMPD1",
      TRUE                          ~ NA_character_
    ),
    Fraction = case_when(
      str_detect(Sample, "_iN_")    ~ "iN",
      str_detect(Sample, "_iDA_")   ~ "iDA",
      TRUE                          ~ NA_character_
    )
  ) %>%
  column_to_rownames("Sample")


# --- 4. plot heatmap, save as pdf and store underlying dataframe as csv 
pdf_file <- file.path(out_dir_diff118_heatmap, "diff118_iN-iDAd35_ASAH1-SMPD1_normCtrl_heatmap.pdf")
cairo_pdf(pdf_file, width = 1.5, height = 5)

pheatmap(
  mat_norm,
  scale              = "row",
  color              = colorRampPalette(rev(brewer.pal(11, "RdYlBu")))(100),
  annotation_col     = annotation_col,
  annotation_row     = row_annotation,
  annotation_colors  = ann_colors,
  breaks             = seq(-2, 2, length.out = 101),
  cluster_rows       = FALSE,
  cluster_cols       = TRUE,
  show_rownames      = FALSE,
  show_colnames      = TRUE,
  angle_col          = 90,
  main               = "diff118 day 23 iN & iDA whole cell proteome: ASAH1/Ctrl, SMPD1/Ctrl",
  fontsize           = 6,
  border_color       = NA
)

dev.off()

# Add cluster as first column to expression matrix
mat_with_cluster <- cbind(cluster = row_annotation$cluster, mat)

# Write to CSV with cluster
write.csv(mat_with_cluster,
          file = file.path(out_dir_diff118_heatmap,"diff118_iN-iDAd35_ASAH1-SMPD1_normCtrl_heatmap.pdf"),
          row.names = TRUE)
```


### IIc.  Violin Plot of Cluster Abundance
### IId. Correlation of Selected Annotations
### III. Scatterplots of selected annotations 
### IV. Barplots of sig. up / down proteins per annotations (used defined)

"log2FC_Ctrl_d23_iDA.Ctrl_d23_iN", "log2FC_ASAH1_d23_iDA.ASAH1_d23_iN", "log2FC_SMPD1_d23_iDA.Ctrl_d23_iDA" 